Workflow Operation "CutMarksToSmil" (#1686)
opencast/opencast: docs/guides/admin/docs/workflowoperationhandlers/cut-marks-to-smil-woh.md

> CutMarksToSmilWorkflowOperationHandler

Reviewer: This is missing from the [list of workflow operation handlers](https://docs.opencast.org/develop/admin/#workflowoperationhandlers/#default-workflow-operations)

This operation parses a JSON containing cut marks into a SMIL that can be used by the 
[VideoEditorWorkflowOperation](editor-woh.md). It does this by attributing the given times to the tracks in the 
given presentation and presenter flavors. 

## Parameter Table

|configuration keys         |example                |description                                                    |
|------------------         |-----------            |---------------------------------------------------------------|
|source-media-flavors       |`presenter/prepared`   |The flavors containing the video tracks. Each flavor must contain at most one file.                   
|source-json-flavor         |`smil/times`           |The flavor of the JSON. Must contain exactly one file.|
|target-smil-flavor         |`smil/cutmarks`        |The flavor of the resulting SMIL.|
|target-tags                |`archive`              |(Optional) Tags to add to the resulting SMIL. Default is `null`|

## JSON Format
The JSON structure specifies all segments which should be kept after cutting.
The property `begin` marks the start of a segment while `duration` its duration.
The values are specified in milliseconds.

>     [
>       {
>         "begin": 1672,
>         "duration": 7199
>       }
>     ]

Reviewer: ```suggestion
```json
[
  {
    "begin": 1672,
    "duration": 7199
  }
]
```
```

|source-media-flavors  |`presenter/prepared` |The flavors containing the video tracks.                       |
|source-json-flavor    |`smil/times`         |The flavor of the JSON. Must contain exactly one file.         |
|target-smil-flavor    |`smil/cutmarks`      |The flavor of the resulting SMIL.                              |
|target-tags           |`archive`            |(Optional) Tags to add to the resulting SMIL. Default is `null`|

JSON Format
-----------

The JSON structure specifies all segments which should be kept after cutting.
The property `begin` marks the start of a segment while `duration` its duration.
The values are specified in milliseconds.

```json
[
  {
    "begin": 1672,
    "duration": 7199
  }
]
```

Operation Example
-----------------

```xml
<operation
    id="cut-marks-to-smil"
    description="Process ingested cutmarks by applying them to current tracks"
    fail-on-error="true"
    exception-handler-workflow="partial-error">
  <configurations>
    <configuration key="source-media-flavors">presenter/prepared,presentation/prepared</configuration>

>     <configuration key="source-json-flavor">json/times</configuration>

Reviewer: I'm wondering if something like `cut-marks/json` wouldn't be a better default name.

Author: "cut-marks" instead of "times" seems more specific and therefore clearer to understand, so +1 for that. As for the order, I put the file type first to be consistent with naming of SMIL catalogue flavors (e.g. `smil/source+partial`), but I got no hard opinion on this.

Reviewer: I decided to try `cut-marks/json`. The `smil/*` is a good argument, but I don't think that smil is really meant as a file type here since it's something specific and not a general purpose file type like JSON. That would make the main type otherwise very unspecific and a wildcard like `json/*` could catch all kinds of things.

 At the same time, we also have other favors with file format as sub-type:
 
- `security/xacml(+…)`
- `captions/google-speech-json`
- `captions/ibm-watson-json`

In the end, it's just the documentation, and you can pick whatever you want anyway. Take a look, and we can still change it if you want to. I'm not that fixed on this. I will push the update shortly.

CutMarksToSmilWorkflowOperationHandler
===================================

Description
-----------

This operation parses a JSON containing cut marks into a SMIL that can be used by the 
[VideoEditorWorkflowOperation](editor-woh.md). It does this by attributing the given times to the tracks in the 
given presentation and presenter flavors. 

## Parameter Table

|configuration keys         |example                |description                                                    |
|------------------         |-----------            |---------------------------------------------------------------|
|source-presenter-flavor    |`presenter/prepared`   |The flavor of the presenter video track. Must contain exactly one file.                   |

> |source-presentation-flavor |`presentation/prepared`|The flavor of the presentation video track. Must contain exactly one file.                  |

Reviewer: Can we not hard-code presentation and presenter but just use `source-flavors` where you can define a list of flavors in the form `a/x,b/x` or a wildcard like `*/source`? That should be negligently harder to do and makes this far more flexible.

Author: I changed it to support a list of flavors (`a/x,b/x`). Does not support wildcards, because I could not figure out how to derive a list of flavors from them. Which I would like to do to check for the amount of tracks in each flavor, because the WHO does not know how it should handle multiple tracks per flavor.

> |source-smil-flavor         |`smil/times`           |The flavor of the JSON. Must contain exactly one file.|

Reviewer 2: Naming this configuration key `source-smil-flavor` is confusing. Based on this name I would expect a smil file, which is xml. However, you expect a JSON with the cut marks from BBB. Therefore I suggest to name it `source-json-flavor` and to amend the description with the information that these are the cut marks from BBB.

Author: Yup, that makes sense.
I would like to keep BBB out of the description, because I wouldn't want people to think that this WHO can only be used with BBB.

|target-smil-flavor         |`smil/cutmarks`        |The flavor of the resulting SMIL.|

## JSON Format

> *begin* marks the start of a segment, *duration* its duration. Times are in milliseconds.

Reviewer: ```suggestion
The JSON structure specifies all segments which should be kept after cutting.
The property `begin` marks the start of a segmen while `duration` its duration.
The values are specified in milliseconds.
```

    [
      {
        "begin": 1672,
        "duration": 7199
      }
    ]

## Operation Example

        <operation
            id="cut-marks-to-smil"
            description="Process ingested cutmarks by applying them to current tracks"
            fail-on-error="true"
            exception-handler-workflow="partial-error">
          <configurations>
            <configuration key="source-presenter-flavor">presenter/prepared</configuration>
            <configuration key="source-presentation-flavor">presentation/prepared</configuration>
            <configuration key="source-json-flavor">smil/times</configuration>
            <configuration key="target-smil-flavor">smil/cutmarks</configuration>
          </configurations>

>         </operation>

Reviewer: - Keep the example as simple as possible
- Directly generate `smil/cutting`
- Use an XML code block

```suggestion
```xml
<operation
    id="cut-marks-to-smil"
    description="Process ingested cutmarks by applying them to current tracks">
  <configurations>
    <configuration key="source-presenter-flavor">presenter/prepared</configuration>
    <configuration key="source-presentation-flavor">presentation/prepared</configuration>
    <configuration key="source-json-flavor">smil/times</configuration>
    <configuration key="target-smil-flavor">smil/cutting</configuration>
  </configurations>
</operation>
```
```

        <operation
            id="clone"
            exception-handler-workflow="partial-error">
          <configurations>
            <configuration key="source-flavor">smil/cutmarks</configuration>
            <configuration key="target-flavor">smil/cutting</configuration>
          </configurations>
        </operation>

        <operation
            id="editor"
            fail-on-error="true"
            exception-handler-workflow="error"
            description="Waiting for user to review / video edit recording">
          <configurations>
            <configuration key="source-flavors">*/prepared</configuration>
            <configuration key="skipped-flavors">*/prepared</configuration>
            <configuration key="smil-flavors">smil/cutting</configuration>
            <configuration key="target-smil-flavor">smil/cutting</configuration>
            <configuration key="target-flavor-subtype">trimmed</configuration>
          </configurations>

>         </operation>

Reviewer: Not sure if we need half a workflow here
```suggestion
```
